### ISSUE TITLE
feat: entropy-weighted credit assignment (GTPO-style)

### ISSUE THREAD

**Issue by teilomillet** [teilomillet/textpolicy#10]
## Parent Issue
Part of #5 (GRPO optimizations for reasoning emergence) — Phase 5: Emergence Acceleration

## Context

This is the **simpler proxy** for HICRA. Instead of mining strategic grams, use entropy as a proxy:

> High entropy token ≈ uncertain ≈ decision point ≈ planning-ish

If this works well enough, full HICRA (#11) may not be needed.

## Problem

Current implementation assigns the same advantage to all tokens:
```python
advantages = rewards_tensor - group_mean  # Same for all tokens
```

This ignores that some tokens are **critical decision points** while others are routine execution.

## Research Evidence

GTPO (2025) shows:
> "High-entropy regions often correspond to steps where the model engages in meaningful exploration and critical logical reasoning."

**Source:** [GTPO and GRPO-S: Token and Sequence-Level Reward Shaping with Policy Entropy](https://arxiv.org/abs/2508.04349)

## The GTPO Formula

```
Â_GTPO(i,t) = Â(i,t) × (1 + β × entropy_normalized(t))

Where:
- Â(i,t) = base advantage (reward - mean)
- β = entropy weight (default 0.1)
- entropy_normalized = token entropy / mean entropy
```

## Proposed Implementation

```python
def compute_token_entropy(logits: mx.array) -> mx.array:
    """Compute entropy at each token position."""
    probs = mx.softmax(logits, axis=-1)
    entropy = -mx.sum(probs * mx.log(probs + 1e-10), axis=-1)
    return entropy


def compute_advantages_gtpo(
    rewards: mx.array,
    token_entropies: mx.array,
    entropy_weight: float = 0.1
) -> mx.array:
    """GTPO: Entropy-weighted advantages."""
    base_advantages = rewards - mx.mean(rewards)
    
    # Normalize entropy within sequence
    entropy_mean = mx.mean(token_entropies, axis=-1, keepdims=True)
    entropy_normalized = token_entropies / (entropy_mean + 1e-8)
    
    # CRITICAL: Detach from gradient to prevent gaming
    entropy_weights = mx.stop_gradient(1 + entropy_weight * (entropy_normalized - 1))
    
    # Weight advantages
    token_advantages = base_advantages[:, None] * entropy_weights
    
    return token_advantages
```

## Litmus Test

| Test | Pass Condition |
|------|----------------|
| **Faster emergence** | Planning language appears earlier than baseline |
| **Higher planning ratio** | > 5% planning tokens by step 500 |
| **Entropy correlation** | High-reward completions have higher entropy at decision points |

## Threshold

- Emergence step with GTPO < emergence step with baseline
- Planning token ratio > 1.5x baseline at step 500
- No accuracy regression

## GTPO vs HICRA Comparison

| Aspect | GTPO (this) | HICRA (#11) |
|--------|-------------|-------------|
| Complexity | Low | High |
| Requires corpus | No | Yes |
| Accuracy | Approximate | Precise |
| Implementation | Hours | Days |

**Recommendation:** Implement this first. If insufficient, add full HICRA.

## Dependencies

- #14 Baseline experiment (to compare against)

## Acceptance Criteria

- [ ] `compute_token_entropy()` function added
- [ ] `compute_advantages_gtpo()` function added
- [ ] Integration with `grpo_loss()` (optional flag)
- [ ] Entropy detached from gradient (`mx.stop_gradient`)
- [ ] Configurable `entropy_weight` parameter
- [ ] A/B comparison vs baseline documented
